text/3720-labeled-match.md

```diff
@@ -0,0 +1,1452 @@
+
+## Edge cases
+
+Overall, behavior is consistent with `loop { match scrutinee { ... } }`
+
+### labels
+
+Labels work as expected for `loop`:
+
+```rust
+fn loop_match() -> Option<u8> {
+    'label: loop match 1u8 {
+        1 => continue 'label 2,
+        2 => continue 'label 3,
+        3 => break 'label Some(42),
+        _ => None
+    }
+}
+```
+
+### `continue <operand>` only in `loop match`
+
+The `const? continue <operand>` expression is only allowed within `loop match` expressions. Usage elsewhere errors, analogous to `break` with value from `for` and `while` loops:
+
+```rust
+for i in 0..10 {
+    continue 42;
+}
+```
+
+```
+error[E0571]: `continue` with value from a `for` loop
+ --> src/main.rs:3:9
+  |
+2 |     for i in 0..10 {
+  |     -------------- you can't `continue` with a value in a `for` loop
+3 |         continue 53i64;
+  |         ^^^^^^^^^^^^^^ can only continue with a value inside `loop match`
+  |
+help: use `continue` on its own without a value inside this `for` loop
+  |
+3 |         continue;
+  |         ~~~~~~~~
+```
+
+### no ambiguity
+
+If unlabeled, `continue <operand>` and `const continue <operand>` continue the innermost loop.
+
+If it is unclear what the user intended when the innermost loop is not a `loop match` (but one of `loop`, `for`, `while`), an error is emitted. We assume that they did in fact intend to continue a `loop match` and emit an error that is analogous to how unlabeled breaks are not allowed in labeled blocks:
+
+```rust
+loop match () {
+    () => {
+        for i in 0..10 {
+            continue ();
+        }
+    }
+}
+```
+
+```
+error[E0000]: `continue` with value from a nested loop
+ --> src/main.rs:3:9
+  |
+4 |         continue ();
+  |         ^^^^^^^^^^^ `continue` with value in a nested loop must bear a label
+  |
+```
+
+### `const continue` must know where to jump
+
+It must be known at compile time which branch the operand of a `loop const match <operand>` or `const continue <operand>` will jump to.
+The rules are described [here](#restrictions-on-the-const-continue-operand).
+If the value is not of the right form, it will be rejected:
+
+```rust
+loop match 1u8 {
+    0 => break,
+    _ => const continue core::hint::black_box(42),
+}
+```
+
+```
+error[E0000]: `continue` with value target unknown at compile time
+ --> src/main.rs:3:9
+  |
+4 |         const continue core::hint::black_box(42),
+  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^ this target of a `continue` is not known at compile time
+  |
+help: use a non-const `continue` with value instead
+  |
+4 |         continue core::hint::black_box(42),
+  |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+### plain `continue` in a `loop match`
+
+A plain `continue` jumps back to the top of the loop like a standard `loop { match _ { ... } }`, so this is accepted:
+
+```rust
+loop match read_next_byte() {
+    0 => continue 1,
+    _ => continue,
```

Comment:
I find this fairly surprising. If I think of `loop match ...` as just a shortcut for `loop { match ... }`, then it makes sense in isolation. But when it's mixed with the `continue value` form, it means the `read_next_byte()` part is only executed in some subset of loop iterations, depending on which state transitions are taken. As a reader I have to (1) remember this every time I see a plain `continue`, and (2) scroll back up to the start of the `loop match` to see the code that runs to determine the next state.

If plain `continue`s were disallowed in `loop match`, I could instead read `loop match read_next_byte()` as: now we're entering a state machine, the initial state is determined by `read_next_byte()`, and to understand the state machine I just chase the transitions from one state to the next. This seems conceptually simpler to me, and doesn't fundamentally lose any expressive power, does it?